File: frontend/src/lib/schemas/planning.ts
/**
 * Zod schemas for the Planning (Planejamento) feature — Phase 5.
 *
 * Mirrors `backend/src/schemas/planning.py`. Planning is a **pure computed
 * service** over the existing fulfillment domain (orders + order_items) plus the
 * WIP tiers. It surfaces two kinds of production suggestions — *cortes* (cutting,
 * grouped by `spec + color_code` with a per-size grade) and *impressões*
 * (printing, one per print design, FRONT as the reference side) — each driven by
 * a dual engine (open-order demand + min-stock reorder), minus work already in
 * production (WIP).
 *
 * The two create endpoints take a list of selected suggestion *keys*; the server
 * recomputes the suggestions inside the transaction and creates PENDING cutting /
 * print orders with no roll / paper assigned yet. Partial success is allowed — a
 * skipped suggestion (stale / silkscreen / no variation / …) never rolls back the
 * created ones.
 */

import { z } from "zod";
import { printDesignRefSchema } from "@/lib/schemas/print-order";
import { SIZES } from "@/lib/schemas/product";

const sizeSchema = z.enum(SIZES);

// The backend `ProductType` enum carries garment names (camiseta, moletom, …),
// which the frontend product schema does not yet model 1:1 (a Phase-1 carry-over
// gap). `product_type` here is purely decorative — it drives the garment glyph —
// so we accept it as a loose string and map it to a glyph at the render site
// (see `garmentGlyphType`), rather than a strict enum that would reject a valid
// response.
const productTypeSchema = z.string();

// String classification fields (plain literals — no PG enum involved, matches the
// prototype wire shape verbatim).
export const PLANNING_STATES = ["pronto", "lisa", "impresso", "ambos"] as const;
export type PlanningState = (typeof PLANNING_STATES)[number];
export const planningStateSchema = z.enum(PLANNING_STATES);

export const SUGGESTION_SOURCES = ["demanda", "estoque"] as const;
export type SuggestionSource = (typeof SUGGESTION_SOURCES)[number];
export const suggestionSourceSchema = z.enum(SUGGESTION_SOURCES);

export const PNG_FLAGS = ["ok", "pending"] as const;
export type PngFlagValue = (typeof PNG_FLAGS)[number];
export const pngFlagSchema = z.enum(PNG_FLAGS);

// ---------- shared refs ----------

export const planningSpecRefSchema = z.object({
  id: z.string(),
  code: z.string(),
  name: z.string(),
});
export type PlanningSpecRef = z.infer<typeof planningSpecRefSchema>;

// ---------- per-demand-SKU breakdown ----------

export const planningSkuSchema = z.object({
  key: z.string(),
  design: printDesignRefSchema,
  spec: planningSpecRefSchema,
  product_type: productTypeSchema,
  color: z.string(),
  color_code: z.string(),
  size: sizeSchema,
  needed: z.number().int(),
  finished: z.number().int(),
  net: z.number().int(),
  blank_have: z.number().int(),
  printed_have: z.number().int(),
  blank_short: z.number().int(),
  printed_short: z.number().int(),
  buildable: z.number().int(),
  state: planningStateSchema,
  order_count: z.number().int(),
});
export type PlanningSku = z.infer<typeof planningSkuSchema>;

// ---------- corte suggestions ----------

export const planningCorteGradeRowSchema = z.object({
  size: sizeSchema,
  qty: z.number().int(),
  demand_qty: z.number().int(),
  stock_qty: z.number().int(),
});
export type PlanningCorteGradeRow = z.infer<typeof planningCorteGradeRowSchema>;

export const planningCorteSchema = z.object({
  key: z.string(),
  spec: planningSpecRefSchema,
  product_type: productTypeSchema,
  color: z.string(),
  color_code: z.string(),
  total: z.number().int(),
  demand: z.number().int(),
  stock: z.number().int(),
  order_count: z.number().int(),
  grade_rows: z.array(planningCorteGradeRowSchema),
  sources: z.array(suggestionSourceSchema),
});
export type PlanningCorte = z.infer<typeof planningCorteSchema>;

// ---------- impressão suggestions ----------

export const planningImpressaoSchema = z.object({
  key: z.string(),
  design: printDesignRefSchema,
  total: z.number().int(),
  demand: z.number().int(),
  stock: z.number().int(),
  order_count: z.number().int(),
  png: pngFlagSchema,
  sources: z.array(suggestionSourceSchema),
});
export type PlanningImpressao = z.infer<typeof planningImpressaoSchema>;

// ---------- totals ----------

export const planningTotalsSchema = z.object({
  toCut: z.number().int(),
  toPrint: z.number().int(),
  cortes: z.number().int(),
  impressoes: z.number().int(),
  demandDriven: z.number().int(),
  stockDriven: z.number().int(),
});
export type PlanningTotals = z.infer<typeof planningTotalsSchema>;

export const planningSuggestionsSchema = z.object({
  skus: z.array(planningSkuSchema),
  cortes: z.array(planningCorteSchema),
  impressoes: z.array(planningImpressaoSchema),
  totals: planningTotalsSchema,
});
export type PlanningSuggestions = z.infer<typeof planningSuggestionsSchema>;

// ---------- bulk-create: cutting ----------

export const planningCutCreatedSchema = z.object({
  key: z.string(),
  cutting_order_id: z.string(),
  code: z.string(),
  total: z.number().int(),
});
export type PlanningCutCreated = z.infer<typeof planningCutCreatedSchema>;

export const planningSkippedSchema = z.object({
  key: z.string(),
  reason: z.enum(["stale", "spec_not_found"]),
});
export type PlanningSkipped = z.infer<typeof planningSkippedSchema>;

export const planningCutResultSchema = z.object({
  created: z.array(planningCutCreatedSchema),
  skipped: z.array(planningSkippedSchema),
  created_count: z.number().int(),
});
export type PlanningCutResult = z.infer<typeof planningCutResultSchema>;

export type PlanningCutCreate = {
  keys: string[];
};

// ---------- bulk-create: printing ----------

export const planningPrintCreatedSchema = z.object({
  key: z.string(),
  print_order_id: z.string(),
  code: z.string(),
  total: z.number().int(),
});
export type PlanningPrintCreated = z.infer<typeof planningPrintCreatedSchema>;

export const planningPrintSkippedSchema = z.object({
  key: z.string(),
  reason: z.enum(["stale", "no_variation", "no_front_side", "silkscreen"]),
});
export type PlanningPrintSkipped = z.infer<typeof planningPrintSkippedSchema>;

export const planningPrintResultSchema = z.object({
  created: z.array(planningPrintCreatedSchema),
  skipped: z.array(planningPrintSkippedSchema),
  created_count: z.number().int(),
});
export type PlanningPrintResult = z.infer<typeof planningPrintResultSchema>;

export type PlanningPrintCreate = {
  keys: string[];
};

File: backend/src/schemas/planning.py
"""Pydantic schemas for the Planning (Planejamento) feature — Phase 5.

Planning is a **pure computed service** over the existing fulfillment domain
(orders + order_items) plus the WIP tiers. It surfaces two kinds of production
suggestions — *cortes* (cutting, grouped by ``spec + color_code`` with a per-size
grade) and *impressões* (printing, one per print design, FRONT as the reference
side) — each driven by a dual engine (open-order demand + min-stock reorder),
minus work already in production (WIP).

The two create endpoints take a list of selected suggestion *keys*; the server
recomputes the suggestions inside the transaction and creates **PENDING** cutting
/ print orders with no roll / paper assigned yet (the operator finishes them on
the Corte / Impressão boards). Partial success is allowed — a skipped suggestion
(stale / silkscreen / no variation / …) never rolls back the created ones.

Wire shapes mirror the prototype ``window.OrionDemand.build`` output verbatim;
see ``services.planning`` for the algorithm.
"""

from __future__ import annotations

import uuid
from typing import Literal

from pydantic import BaseModel, Field

from models.enums import ProductType, Size
from schemas.print_order import PrintDesignRef

# String enums for the computed classification fields (plain literals so the
# wire shape matches the prototype exactly and no PG enum is involved).
PlanningState = Literal["pronto", "lisa", "impresso", "ambos"]
SuggestionSource = Literal["demanda", "estoque"]
PngFlag = Literal["ok", "pending"]


# ---------- shared refs ----------


class PlanningSpecRef(BaseModel):
    """Minimal product-spec (ficha técnica) projection embedded in suggestions."""

    id: uuid.UUID
    code: str
    name: str


# ---------- per-demand-SKU breakdown ----------


class PlanningSku(BaseModel):
    """One open-demand SKU ``(design, spec, color_code, size)`` with component availability.

    ``needed`` counts open ``OrderItem`` rows; ``net = max(0, needed - finished)``
    is the quantity that still has to be assembled. ``blank_short`` / ``printed_short``
    say which component is missing; ``buildable = min(net, blank_have, printed_have)``.
    """

    key: str
    design: PrintDesignRef
    spec: PlanningSpecRef
    product_type: ProductType
    color: str
    color_code: str
    size: Size
    needed: int
    finished: int
    net: int
    blank_have: int
    printed_have: int
    blank_short: int
    printed_short: int
    buildable: int
    state: PlanningState
    order_count: int


# ---------- corte suggestions ----------


class PlanningCorteGradeRow(BaseModel):
    """Per-size piece count of a corte suggestion (split into demand vs stock)."""

    size: Size
    qty: int
    demand_qty: int
    stock_qty: int


class PlanningCorte(BaseModel):
    """A suggested cutting order, grouped by ``spec + color_code`` with a per-size grade."""

    key: str  # "{spec_id}|{color_code}"
    spec: PlanningSpecRef
    product_type: ProductType
    color: str
    color_code: str
    total: int
    demand: int
    stock: int
    order_count: int
    grade_rows: list[PlanningCorteGradeRow]
    sources: list[SuggestionSource]


# ---------- impressão suggestions ----------


class PlanningImpressao(BaseModel):
    """A suggested print order, one per print design (FRONT as the reference side)."""

    key: str  # "{print_design_id}"
    design: PrintDesignRef
    total: int
    demand: int
    stock: int
    order_count: int
    png: PngFlag
    sources: list[SuggestionSource]


# ---------- totals ----------


class PlanningTotals(BaseModel):
    toCut: int  # noqa: N815 — wire shape (mirrors the prototype + frontend)
    toPrint: int  # noqa: N815 — wire shape
    cortes: int
    impressoes: int
    demandDriven: int  # noqa: N815 — wire shape
    stockDriven: int  # noqa: N815 — wire shape


class PlanningSuggestions(BaseModel):
    """The whole computed model — returned bare (like ``dashboard/summary``)."""

    skus: list[PlanningSku]
    cortes: list[PlanningCorte]
    impressoes: list[PlanningImpressao]
    totals: PlanningTotals


# ---------- bulk-create: cutting ----------


class PlanningCutCreate(BaseModel):
    keys: list[str] = Field(min_length=1)


class PlanningCutCreated(BaseModel):
    key: str
    cutting_order_id: uuid.UUID
    code: str
    total: int


class PlanningSkipped(BaseModel):
    key: str
    reason: Literal["stale", "spec_not_found"]


class PlanningCutResult(BaseModel):
    created: list[PlanningCutCreated]
    skipped: list[PlanningSkipped]
    created_count: int


# ---------- bulk-create: printing ----------


class PlanningPrintCreate(BaseModel):
    keys: list[str] = Field(min_length=1)


class PlanningPrintCreated(BaseModel):
    key: str
    print_order_id: uuid.UUID
    code: str
    total: int


class PlanningPrintSkipped(BaseModel):
    key: str
    reason: Literal["stale", "no_variation", "no_front_side", "silkscreen"]


class PlanningPrintResult(BaseModel):
    created: list[PlanningPrintCreated]
    skipped: list[PlanningPrintSkipped]
    created_count: int


__all__ = [
    "PlanningCorte",
    "PlanningCorteGradeRow",
    "PlanningCutCreate",
    "PlanningCutCreated",
    "PlanningCutResult",
    "PlanningImpressao",
    "PlanningPrintCreate",
    "PlanningPrintCreated",
    "PlanningPrintResult",
    "PlanningPrintSkipped",
    "PlanningSkipped",
    "PlanningSku",
    "PlanningSpecRef",
    "PlanningSuggestions",
    "PlanningTotals",
]
